### ISSUE TITLE
global.json sdk preference doesn't appear to be respected

### ISSUE THREAD

**Issue by davkean** [dotnet/cli#9721]
_From @flcdrg on July 22, 2018 12:1_

I have a global.json with the following content:

```
{
    "sdk": {
      "version": "2.1.300"
    }
}
```

And a project with the following header:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net471</TargetFramework>
```

On my system, running `dotnet --info` gives the following output:

```
.NET Core SDK (reflecting any global.json):
 Version:   2.1.302
 Commit:    9048955601

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.302\

Host (useful for support):
  Version: 2.1.2
  Commit:  811c3ce6c0

.NET Core SDKs installed:
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.100-preview-007326 [C:\Program Files\dotnet\sdk]
  2.1.100-preview-007354 [C:\Program Files\dotnet\sdk]
  2.1.100-preview-007363 [C:\Program Files\dotnet\sdk]
  2.1.100 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.200-preview-007474 [C:\Program Files\dotnet\sdk]
  2.1.200-preview-007517 [C:\Program Files\dotnet\sdk]
  2.1.200-preview-007576 [C:\Program Files\dotnet\sdk]
  2.1.200-preview-007597 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.302 [C:\Program Files\dotnet\sdk]
  2.1.400-preview-009063 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

But when I run msbuild, the following is displayed:

```
Microsoft (R) Build Engine version 15.7.180.61344 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\2.1.400-preview-009063\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInfere
nce.targets(143,5): message NETSDK1057: You are working with a preview version of the .NET Core SDK. You can define the
 SDK version via a global.json file in the current project. More at https://go.microsoft.com/fwlink/?linkid=869452
```

Why is it saying it's using the preview version of .NET Core SDK if global.json specifies 2.1.300?

_Copied from original issue: dotnet/project-system#3769_

**Comment by peterhuene** [comment]
Hi @flcdrg.  Are you building your project with `msbuild.exe` or using `dotnet build`?

**Comment by flcdrg** [comment]
msbuild.exe at the moment

**Comment by peterhuene** [comment]
@nguerrera Is this something that is supposed to work when invoking MSBuild.exe directly?

**Comment by nguerrera** [comment]
Yes, it is supposed to work.

Is global.json at the solution file level (or above) or only at the project file level (or above). You need to put it at the solution level (or above) for VS / msbuild.exe sdk resolver to use it (unless you're msbuilding a csproj with no sln).

If it's not that, can you try the following:

```
set COREHOST_TRACE=1
msbuild 2> hostfxr_log.txt
```

Then share hostfxr.log?

**Comment by nguerrera** [comment]
Another note that is possibly more relevant than sln vs csproj.

`dotnet` uses the current working directory or above
`msbuild` uses the solution directory or above,

**Comment by flcdrg** [comment]
Global.json is at the solution level

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by nguerrera** [comment]
OK, can you  try the commands above and send the log.

```
set COREHOST_TRACE=1
msbuild 2> hostfxr_log.txt
```

**Comment by flcdrg** [comment]
Tracing enabled

--- Invoked hostfxr [commit hash: 1632fa1589b0eee3277a8841ce1770e554ece037] hostfxr_resolve_sdk

Probing path [C:\dev\work\Example\SVHA.sln\global.json] for global.json

Probing path [C:\dev\work\Example\global.json] for global.json

Found global.json [C:\dev\work\Example\global.json]

--- Resolving CLI version from global json [C:\dev\work\Example\global.json]

CLI version is [2.1.300] in global json file [C:\dev\work\Example\global.json]

Searching SDK directory in [C:\Program Files\dotnet]

--- Resolving SDK version from SDK dir [C:\Program Files\dotnet\sdk]

Considering version... [.]

Considering version... [..]

Considering version... [2.1.100]

Considering version... [2.1.100-preview-007326]

Considering version... [2.1.100-preview-007354]

Considering version... [2.1.100-preview-007363]

Considering version... [2.1.101]

Considering version... [2.1.103]

Considering version... [2.1.104]

Considering version... [2.1.200]

Considering version... [2.1.200-preview-007474]

Considering version... [2.1.200-preview-007517]

Considering version... [2.1.200-preview-007576]

Considering version... [2.1.200-preview-007597]

Considering version... [2.1.201]

Considering version... [2.1.202]

Considering version... [2.1.302]

Considering version... [2.1.4]

Considering version... [2.1.400-preview-009063]

Considering version... [NuGetFallbackFolder]

Checking if resolved SDK dir [C:\Program Files\dotnet\sdk\2.1.400-preview-009063] exists

Resolved SDK dir is [C:\Program Files\dotnet\sdk\2.1.400-preview-009063]

Searching SDK directory in [C:\Program Files (x86)\dotnet]

--- Resolving SDK version from SDK dir [C:\Program Files (x86)\dotnet\sdk]

Considering version... [.]

Considering version... [..]

Considering version... [2.1.302]

Considering version... [NuGetFallbackFolder]

Checking if resolved SDK dir [C:\Program Files (x86)\dotnet\sdk\2.1.302] exists

Resolved SDK dir is [C:\Program Files (x86)\dotnet\sdk\2.1.302]

Found CLI SDK in: C:\Program Files\dotnet\sdk\2.1.400-preview-009063

**Comment by nguerrera** [comment]
Thanks. This looks like an issue fixed by https://github.com/dotnet/core-setup/pull/3711, but the copy of the dll used by VS  and msbuild.exe won't get the fix until VS 15.8. You should get the same SDK resolved from VS 15.8 msbuild.exe and `dotnet`.

In the meantime, your best workaround is to specify a version of global.json that is installed on your machine. In this case, put 2.1.302 in global.json in place of 2.1.300 that is not installed. 

The behavior difference kicks in when no exact match is found and it is deciding on the next best thing to use.

**Comment by livarcocc** [comment]
Closing as fixed.
